pkg/vm/vm.go

```diff
@@ -22,37 +28,302 @@ type VirtualMachineConfig struct {
+	}
+
+	vm.Step += 1
 	return nil
 }
+func (vm *VirtualMachine) RunStepAt(pc uint64) error {
+	vm.Context.Pc = pc
+	return vm.RunStep()
+}
 
 func (vm *VirtualMachine) RunInstruction(instruction *Instruction) error {
+	// todo(rodro): any OffOpX can be negative, a better math system is required due to
+	// substraction. Also it will need to handle overflows and underflows
+	dstCell, err := vm.getCellDst(instruction)
+	if err != nil {
+		return err
+	}
+
+	op0Cell, err := vm.getCellOp0(instruction)
+	if err != nil {
+		return err
+	}
+
+	op1Cell, err := vm.getCellOp1(instruction, op0Cell)
+	if err != nil {
+		return err
+	}
+
+	res, err := vm.computeRes(instruction, op0Cell, op1Cell)
+	if err != nil {
+		return err
+	}
+
+	err = vm.opcodeAssertions(instruction, dstCell, op0Cell, res)
+	if err != nil {
+		return err
+	}
+
+	nextPc, err := vm.updatePc(instruction, dstCell, op1Cell, res)
+	if err != nil {
+		return err
+	}
+
+	nextAp, err := vm.updateAp(instruction, res)
+	if err != nil {
+		return err
+	}
+
+	nextFp, err := vm.updateFp(instruction, res)
+	if err != nil {
+		return err
+	}
+
+	vm.Context.Pc = nextPc
+	vm.Context.Ap = nextAp
+	vm.Context.Fp = nextFp
+
+	return nil
+}
+
+func (vm *VirtualMachine) RunHint() error {
 	return nil
 }
 
-func (vm *VirtualMachine) RunHint(instruction *Instruction) error {
+func (vm *VirtualMachine) getCellDst(instruction *Instruction) (*mem.Cell, error) {
+	var dstRegister uint64
+	if instruction.DstRegister == Ap {
+		dstRegister = vm.Context.Ap
+	} else {
+		dstRegister = vm.Context.Fp
+	}
+
+	// todo(rodro): fix this math
+	return vm.MemoryManager.Memory.Peek(dataSegment, dstRegister+uint64(instruction.OffDest))
+}
+
+func (vm *VirtualMachine) getCellOp0(instruction *Instruction) (*mem.Cell, error) {
+	var op0Register uint64
+	if instruction.Op0Register == Ap {
+		op0Register = vm.Context.Ap
+	} else {
+		op0Register = vm.Context.Fp
+	}
+	// todo(rodro): fix this math
+	offset := op0Register + uint64(instruction.OffOp0)
+	return vm.MemoryManager.Memory.Peek(dataSegment, offset)
+}
+
+func (vm *VirtualMachine) getCellOp1(instruction *Instruction, op0Cell *mem.Cell) (*mem.Cell, error) {
+	var op1Address *mem.MemoryAddress
+	switch instruction.Op1Source {
+	case Op0:
+		// in this case Op0 is being used as an address, and must be of unwrapped as is
+		op0Address, err := op0Cell.Read().ToMemoryAddress()
+		if err != nil {
+			return nil, fmt.Errorf("expected op0 to be an address: %w", err)
+		}
+		op1Address = mem.CreateMemoryAddress(op0Address.SegmentIndex, op0Address.Offset)
+	case Imm:
+		// todo(rodro): would it be sensitive to check instruction.OffOp1 == 1?
+		op1Address = mem.CreateMemoryAddress(programSegment, vm.Context.Pc)
+	case FpPlusOffOp1:
+		op1Address = mem.CreateMemoryAddress(dataSegment, vm.Context.Fp)
+	case ApPlusOffOp1:
+		op1Address = mem.CreateMemoryAddress(dataSegment, vm.Context.Ap)
+	}
+	// todo(rodro): fix this math
+	op1Address.Offset += uint64(instruction.OffOp1)
```

Comment:
I don't think that is the best option, offsets should be always positive. A memory address from a segment shouldn't point to the previous segment. due to a negative offset. I was thinking on the following opperation. Since we know offsets are in [2**-15, 2**15], to make something like:

op1 = op1 + (offOp1 + 2**15) - 2**15

And so Op1 should be always positive

Reply:
I described the operation that way since offOp1 is the only signed integer, that way it can be safely casted to uint64

Reply:
> A memory address from a segment shouldn't point to the previous segment. due to a negative offset.
Why is that? Offsetting cannot change a segment at the level of the execution of the vm offsetting back and forth should keep an address in its segment

Reply:
What I meant is that we shouldn't use signed integers for the offset because it will always be non-negative. For example if we have an address at (s, -1), after relocation it would point actually to the previous segment (if it exists) like (s - 1, t) which wouldn't be right

Reply:
In the original python VM they kept their offsets positive yes, but they were using python ```int``` which is a ```BitInt``` in fact so if an offset went negative they'd normalize it by modding by the prime, e.g. an offset ```-1``` would turn into a memory cell ```p-1```. If we want to stay in the 64 ints (which a cairo program shouldn't probably exceed and a smart contract won't for obvious gas reasons :) ) then we should wrap in a clever way. Firstly, wrappings can happen for sure, consider the instruction ```[fp-10] = 42``` at the beginning of a program (no previous calls). Where should this ```fp - 10``` point at? At the beginning fp points to the data segment first cell. Let's say we relocate the data segment to a felt ```d``` (I mean it starts from that memory address). Fine, then ```d - 10 (mod p)``` should point at ```42```. But what can ```d-10``` be? If we relocated other segments prior to this one then d-10 might point to the previous one, not ideal, then the assertion [fp - 10] = 42 might fail. If d < 10 then we wrap up and get to the far end of the field, which is not ideal because of possible clashes with other segments again (how do we solve conflicts of different segments at the end of the field? :)). The advantage of the solution with negative offsets in a segment is that the segment is kind of contiguous it spans from -least offset to +largest offset which is easily relocatable into spare space of the final field addressed memory

Reply:
For instance if the least offset of a segment is -10 then we can relocate it to MAX_ADDRESSED_ADDR + 10 and we won't interfere with another segment

Reply:
Ok just a couple of things to make sure we are on the same ground. `fp` does not point to the beginning of a segment, it just points at the beginning of the stored variables of the current function being executed. So `fp` is probably never at the beginning of the segments it moves around.

Now, instead of `fp` lets use a generic address located at segment `s` and offset `x` and `d` where `d > x`. Then having a new address such as `(s, x - d)` will indeed point towards a negative offset.

So if we do smth like `assert memory(s, x - d) = 42` then it can have different approaches. As you mentioned we can set that the real beginning of `s` is indeed `x - d`  as you mentioned previously, and relocate all memory values accordingly once the execution is finalized. This solution would then require some technical changes as well in the vm. For example, we would need to somehow represent this negative offset, using arrays alone won't be possible any more and we would probably switch to maps

Now I think this is a very niche case, why would someone like to write "off" memory. There is no I use for it that I can think off. Neither old nor new compiler would produce that kind of code (from what I know), just someone writing Casm that wants on a whim that their segment starts on `x - d`. So the vm specifications in theory doesn't prohibit it, but in practice it has no use that I can think of.

Also I did a quick test on the Cairo playground and writing "off" memory fails there as well with a textless error
```
%builtins output

from starkware.cairo.common.serialize import serialize_word

func main{output_ptr: felt*}() {
    [fp - 10] = 15;
    return ();
}
```

Reply:
I see! You convinced me, thank you! :) Yeah, I think integer overflow check requires less operations than making offsets negative, so yeah let's leave it as is :)